File: capacitor-plugins/capacitor-native/ios/Sources/AonsokuNativePlugin/Data/Database/SearchHelper.swift
import Foundation
import GRDB

enum SearchHelper {
    static func tokenize(_ query: String) -> [String] {
        query.folding(
            options: [.diacriticInsensitive, .caseInsensitive, .widthInsensitive],
            locale: nil
        )
        .split(whereSeparator: \.isWhitespace)
        .map(String.init)
        .filter { !$0.isEmpty }
    }

    static func buildCondition(
        query: String, columns: [String]
    ) -> SQLExpression? {
        let tokens = tokenize(query)
        guard !tokens.isEmpty else { return nil }

        let normalize = DatabaseManager.normalizeFunction
        var allTokenConditions: SQLExpression?

        for token in tokens {
            let pattern = "%\(token)%"
            var columnConditions: SQLExpression?

            for colName in columns {
                let col = Column(colName)
                let condition = normalize(col).like(pattern)
                if let existing = columnConditions {
                    columnConditions = existing || condition
                } else {
                    columnConditions = condition
                }
            }

            if let tokenCondition = columnConditions {
                if let existing = allTokenConditions {
                    allTokenConditions = existing && tokenCondition
                } else {
                    allTokenConditions = tokenCondition
                }
            }
        }

        return allTokenConditions
    }
}

File: capacitor-plugins/capacitor-native/android/src/main/java/github/realtvop/aonsoku/plugins/data/SearchHelper.kt
package github.realtvop.aonsoku.plugins.data
import java.text.Normalizer
object SearchHelper {
    fun normalize(text: String) = Normalizer.normalize(text, Normalizer.Form.NFD).replace(Regex("\\p{M}"), "").lowercase()
    fun tokenize(query: String) = normalize(query).split(Regex("\\s+")).filter { it.isNotEmpty() }
    fun buildCondition(query: String, columns: List<String>): String? {
        val tokens = tokenize(query)
        if (tokens.isEmpty()) return null
        return tokens.joinToString(" AND ") { token ->
            columns.joinToString(" OR ") { col ->
                "LOWER($col) LIKE '%$token%'"
            }.let { "($it)" }
        }
    }
}
